test/plugin/base_test.rb

```diff
@@ -2374,7 +2506,12 @@ def verify_cascading_json_detection_with_log_entry_fields(
         d.emit(input_log_entry)
         d.run
       end
-      verify_log_entries(1, COMPUTE_PARAMS, 'jsonPayload') do |entry|
+      expected_params = COMPUTE_PARAMS.dup
+      if log_entry_field == 'labels'
```

Comment:
Why are we doing this here, rather than in the invocation of this method? It seems like it would be better to define `COMPUTE_PARAMS` without labels and then verify the labels in the supplied block...

Reply:
Unless we refactor the `verify_json_log_entries` method, we could not stop `verify_log_entries` from verifying `labels`.

One thing we could do (in fact I started it like this but decided not to do it) is to add two params: `expected_params_from_root` and `expected_params_from_nested`. We can leave the default value as `COMPUTE_PARAMS`, and pass in customized value when testing `labels`. e.g.:

```
  def test_cascading_json_detection_with_log_entry_labels_field
     verify_cascading_json_detection_with_log_entry_fields(
       'labels', DEFAULT_LABELS_KEY,
       root_level_value: LABELS_MESSAGE,
-      nested_level_value: LABELS_MESSAGE2)
+      nested_level_value: LABELS_MESSAGE2,
+      expected_value_from_root: LABELS_MESSAGE.merge(COMPUTE_PARAMS[:labels]),
+      expected_value_from_nested: LABELS_MESSAGE2.merge(
+        COMPUTE_PARAMS[:labels]),
+      expected_params_from_root: COMPUTE_PARAMS.merge(
+        labels: COMPUTE_PARAMS[:labels].merge(LABELS_MESSAGE)),
+      expected_params_from_nested: COMPUTE_PARAMS.merge(
+        labels: COMPUTE_PARAMS[:labels].merge(LABELS_MESSAGE2)))
   end

   def test_cascading_json_detection_with_log_entry_operation_field
@@ -2460,6 +2467,10 @@ module BaseTest
       :expected_value_from_root, root_level_value)
     expected_value_from_nested = expectation.fetch(
       :expected_value_from_nested, nested_level_value)
+    expected_params_from_root = expectation.fetch(
+      :expected_params_from_root, COMPUTE_PARAMS)
+    expected_params_from_nested = expectation.fetch(
+      :expected_params_from_nested, COMPUTE_PARAMS)

     setup_gce_metadata_stubs

@@ -2498,21 +2509,27 @@ module BaseTest
       default_key => root_level_value)

     {
-      log_entry_with_root_level_field => expected_value_from_root,
-      log_entry_with_nested_level_field => expected_value_from_nested,
-      log_entry_with_both_level_fields => expected_value_from_nested
-    }.each_with_index do |(input_log_entry, expected_value), index|
+      log_entry_with_root_level_field => {
+        expected_value: expected_value_from_root,
+        expected_params: expected_params_from_root
+      },
+      log_entry_with_nested_level_field => {
+        expected_value: expected_value_from_nested,
+        expected_params: expected_params_from_nested
+      },
+      log_entry_with_both_level_fields => {
+        expected_value: expected_value_from_nested,
+        expected_params: expected_params_from_nested
+      }
+    }.each_with_index do |(input_log_entry, expected), index|
+      expected_params = expected[:expected_params]
+      expected_value = expected[:expected_value]
       setup_logging_stubs do
         @logs_sent = []
         d = create_driver(DETECT_JSON_CONFIG)
         d.emit(input_log_entry)
         d.run
       end
-      expected_params = COMPUTE_PARAMS.dup
-      if log_entry_field == 'labels'
-        expected_value = expected_value.merge(COMPUTE_PARAMS[:labels])
-        expected_params[:labels] = expected_value
-      end
```

I chose not to do it because it increased complexity and is kind of distracting. `expected_value` is what we need to focus on. `expected_params` just happens to be something we need to placate.

Does that make sense?

Reply:
Note that I suggested removing labels from `COMPUTE_PARAMS`, which would make the expected and actual params identical. You can then use the block within the verification step to verify the labels separately. WDYT?

Reply:
How would we bypass [this check](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/946798d787bef1ef2e38ec8ec327bdd4ac7e4bce/test/plugin/base_test.rb#L2203) though? The actual params has `LABELS_MESSAGE.merge(COMPUTE_PARAMS[:labels])` as the labels. Removing labels from `COMPUTE_PARAMS` won't help, right? Unless we want to refactor `verify_json_log_entries`.

Reply:
I'm actually leaning towards refactoring `verify_json_log_entries`, because it conflates two checks — the fact that certain labels are present _somewhere_ in the write logs request, and the fact that the individual log entry contains specific labels. One fix would be to change `check_labels` to only verify the existence of the expected labels, rather than full equality of the label set. We might need to add extra checks for the resource labels (to ensure that there's exact equality there). WDYT?

Reply:
Refactored `verify_json_log_entries`. I added a flag `check_exact_entry_labels`. Also refactored `check_labels` a little bit in the end.